_Static_assert failures should be reported in the front-end

tautschnig opened the pull request:
We left it to a failing invariant in goto-program conversion to detect a
failing (local) _Static_assert. Global _Static_assert statements were
not verified at all.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- n/a White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

tautschnig committed:
_Static_assert failures should be reported in the front-end

We left it to a failing invariant in goto-program conversion to detect a
failing (local) _Static_assert. Global _Static_assert statements were
not verified at all.

tautschnig committed:
__builtin_isinf{,_sign} require a floating-point argument

This isn't necessarily spelled out in the signature of the function, but
GCC's documentation states this requirement.

martin-cs reviewed (approved):
OK but I do wonder whether we should treat these like `assert`s.  I guess the point is that they are supposed to be build time and if they fail then the code is explicitly marked as being wrong.

kroening commented on regression/cbmc-library/Float_lib1/main.c:
```diff
@@ -43,10 +43,10 @@ int main()
 
   _Static_assert(
-    __builtin_isinf_sign(0) == 0, "__builtin_isinf_sign is constant");
+    __builtin_isinf_sign(0.0) == 0, "__builtin_isinf_sign is constant");
```
Shouldn't that be fixed rather than changing the test?

tautschnig replied:
Well, it's the test that was broken: running GCC on this test before the change said:
```
$ gcc regression/cbmc-library/Float_lib1/main.c
regression/cbmc-library/Float_lib1/main.c: In function ‘main’:
regression/cbmc-library/Float_lib1/main.c:46:35: error: non-floating-point argument in call to function ‘__builtin_isinf’
   _Static_assert(!__builtin_isinf(0), "__builtin_isinf is constant");
                                   ^
regression/cbmc-library/Float_lib1/main.c:46:18: error: expression in static assertion is not an integer
   _Static_assert(!__builtin_isinf(0), "__builtin_isinf is constant");
                  ^
regression/cbmc-library/Float_lib1/main.c:49:26: error: non-floating-point argument in call to function ‘__builtin_isinf_sign’
     __builtin_isinf_sign(0) == 0, "__builtin_isinf_sign is constant");
                          ^
regression/cbmc-library/Float_lib1/main.c:49:5: error: expression in static assertion is not an integer
     __builtin_isinf_sign(0) == 0, "__builtin_isinf_sign is constant");
     ^~~~~~~~~~~~~~~~~~~~
```

kroening replied:
Ah, ok, gcc/clang also error this.

kroening commented on src/ansi-c/c_typecheck_code.cpp:
```diff
@@ -102,9 +103,24 @@ void c_typecheck_baset::typecheck_code(codet &code)
     typecheck_expr(code.op0());
     typecheck_expr(code.op1());
+
+    code.op0() = typecast_exprt::conditional_cast(code.op0(), bool_typet{});
```
That might not work, consider `implicit_typecast_bool`?

tautschnig replied:
Done!

tautschnig commented:
#5679 needs to be merged first to make CI happy again.

kroening commented on src/util/simplify_expr_floatbv.cpp:
```diff
@@ -20,9 +20,6 @@ Author: Daniel Kroening, kroening@kroening.com
 simplify_exprt::resultt<>
 simplify_exprt::simplify_isinf(const unary_exprt &expr)
 {
-  if(expr.op().type().id() != ID_floatbv)
```
I would leave that as a PRECONDITION.

tautschnig replied:
Could do, but would be redundant with `to_floatbv_type(expr.type())` that `ieee_floatt::from_expr` does. I'm happy to add it for clarity, but if so I'd suggest to add it to `simplify_isnan` and `simplify_isnormal` as well?

kroening replied:
Yes, please add, the idea being "fail early" and "clarity".

tautschnig replied:
Done!

tautschnig committed:
isinf, isnan, isnormal require floatbv operands

Fail early if simplification encounters non-floatbv operands to these
functions. This would have been caught by ieee_floatt::from_expr, but
failing early makes diagnosing easier.
